packages/labs/src/frameworks/react/create-component.ts

```diff
@@ -0,0 +1,216 @@
+
+const reservedReactProperties = new Set([
+  'children',
+  'localName',
+  'ref',
+  // TODO(sorvell): why are the properties below included?
+  'style',
+  'className',
+]);
+
+const listenedEvents: WeakMap<
+  Element,
+  Map<string, EventListenerObject>
+> = new WeakMap();
+
+const setEvent = (
+  node: Element,
+  event: string,
+  value: (event?: Event) => void
+) => {
+  let events = listenedEvents.get(node);
+  if (events === undefined) {
+    listenedEvents.set(node, (events = new Map()));
+  }
+  let handler = events.get(event);
+  if (value !== undefined) {
+    // If necessary, add listener and track handler
+    if (handler === undefined) {
+      events.set(event, (handler = {handleEvent: value}));
+      node.addEventListener(event, handler);
+      // Otherwise just update the listener with new value
+    } else {
+      handler.handleEvent = value;
+    }
+    // Remove listener if one exists and value is undefined
+  } else if (handler !== undefined) {
+    events.delete(event);
+    node.removeEventListener(event, handler);
+  }
+};
+
+const setProperty = <E extends Element, T>(
+  node: E,
+  name: string,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  value: any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  old: any,
+  events?: T
+) => {
+  // Dirty check and prevent setting reserved properties.
+  if (value === old || reservedReactProperties.has(name as string)) {
+    return;
+  }
+  // For events, use an explicit list.
+  const event = (events?.[name as keyof T] as unknown) as string;
+  if (event !== undefined) {
+    setEvent(node, event, value);
+  } else {
+    node[name as keyof E] = value;
+  }
+};
+
+type Events<S> = {
+  [P in keyof S]?: (e: Event) => unknown;
+};
+
+type Constructor<T> = {new (): T};
+
+/**
+ *  Creates a React component from a CustomElement.
+ */
+export const createComponent = <I extends HTMLElement, E>(
+  React: typeof ReactModule,
+  tagName: string,
+  elementClass: Constructor<I>,
+  events?: E
+) => {
+  const Component = React.Component;
+  const createElement = React.createElement;
+
+  // Props the user is allowed to use, includes standard attributes, children,
+  // ref, as well as special event and element properties.
+  type UserProps = React.PropsWithChildren<
+    React.PropsWithRef<Partial<I> & Events<E>>
+  >;
+
+  // Props used by this component wrapper. This is the UserProps and the
+  // special `__forwardedRef` property. Note, this ref is special because
+  // it's both needed in this component to get access to the rendered element
+  // and must fulfill any ref passed by the user.
+  type ComponentProps = UserProps & {
+    __forwardedRef?: React.Ref<unknown>;
+  };
+
+  // List of properties/events which should be specially handled by the wrapper
+  // and not handled directly by React.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const elementPropsMap: {[index: string]: any} = {...(events ?? {})};
```

Comment:
Changed to unknown. Using an object helps type the corresponding `_elementProps` property since it can be `typeof elementPropsMap`. Is that possible using a Set?

Reply:
Fixed.